File: Maximal Score After Applying K Operations.py
class Solution:
    def maxKelements(self, nums: List[int], k: int) -> int:
        nums = [-num for num in nums]
        heapq.heapify(nums)
        score = 0
        while k:
            val = heapq.heappop(nums)
            score -= val
            heapq.heappush(nums, -math.ceil(-val / 3))
            k -= 1
        return score

# class Solution:
#     def maxKelements(self, nums: List[int], k: int) -> int:
#         q = [-num for num in nums]
#         heapq.heapify(q)
#         score = 0
#
#         while k:
#             num = heapq.heappop(q)
#             score -= num
#             heapq.heappush(q, math.floor(num / 3))
#             k -= 1
#
#         return score

File: Maximal Score After Applying K Operations.cpp
class Solution {
public:
    long long maxKelements(vector<int>& nums, int k) {
        make_heap(nums.begin(), nums.end());
        int64_t score = 0;
        while (k) {
            pop_heap(nums.begin(), nums.end());
            int64_t val = nums.back();
            nums.pop_back();
            score += val;
            val = ceil((double) val / 3.);
            nums.push_back(val);
            push_heap(nums.begin(), nums.end());
            k--;
        }
        return score;
    }
};
